cladecombiner/aggregator.py

```diff
@@ -0,0 +1,228 @@
+from abc import ABC, abstractmethod
+from collections.abc import Iterable
+from warnings import warn
+
+from .taxon import Taxon
+from .taxon_utils import sort_taxa
+from .taxonomy_scheme import PhylogeneticTaxonomyScheme
+from .utils import table
+
+
+class Aggregation(dict[Taxon, Taxon]):
+    """
+    An object for aggregations, basically just a dictionary.
+    """
+
+    def _validate(
+        self, input_taxa: Iterable[Taxon], taxon_map: dict[Taxon, Taxon]
+    ):
+        """
+        Checks that all input taxa have been mapped exactly once.
+        """
+        if set(taxon_map.keys()) != set(input_taxa):
+            raise RuntimeError(
+                "Mismatch between aggregated taxa and input taxa. Input taxa are: "
+                + str(input_taxa)
+                + " but aggregated taxa are "
+                + str(taxon_map.keys())
+            )
+        tab = table(taxon_map)
+        if not all(v == 1 for v in tab.values()):
+            raise RuntimeError(
+                "Found following taxa mapped more than once: "
+                + str([k for k, v in tab.items() if v > 1])
+            )
+
+    def __init__(
+        self, input_taxa: Iterable[Taxon], taxon_map: dict[Taxon, Taxon]
+    ):
+        self._validate(input_taxa, taxon_map)
+        super().__init__(taxon_map)
+
+    def to_str(self):
+        """
+        Get str : str map of taxa names
+        """
+        return {k.name: v.name for k, v in self.items()}
+
+
+class Aggregator(ABC):
```

Comment:
I want some minimal validation automatically. Every taxon should be mapped exactly once is the minimum correctness standard, though some `Aggregator`s might know more about what to check.

The choices seemed to be what I did (make the map check) or have `Aggregator.aggregate()` be a non-abstract method calling some actual aggregation function `Aggregator._aggregate()` and then `Aggregator.validate()`. Neither is ideal, but I'm open to being told one is less awkward than the other.

I'm also going to dispute the above "then the taxon map can be just a map, with no further functionality." Exporting the `Taxon : Taxon` dictto a `str : str` dict is going to be one of the most common usage patterns for the existence of this class, so it might as well be a method. `Taxon` objects are our internal representations, but everywhere else a user is going to need to interact with taxa, they'll be strings.